### DISCUSSION TITLE
unknown error message for point_stat

### DISCUSSION THREAD

**Discussion by Ho-ChunHuang-NOAA** [dtcenter/METplus#1496] [Statistical Computation] [answered]
I am running aqm daily max/ave for o3/and pm25.  The error message appeared for OZMAX1 and OZMAX8 point_stat but not for PMMAX and PMAVE. 

If the valid day is 03/06/2022, the point-stat for the OZMAX1 and OZMAX8 on 03/05 and 03/04 finished successfully but failed on 03/03.  The intermediated log indicated it can not find matched OZMAX1/L1 and OZMAX8/L1 from model output grib2.  However, $WGRIB2 show it did have the variable.  

If I set valid day is 03/05/2022, the point-stat for the OZMAX1 and OZMAX8 on 03/04 and 03/03 finished successfully, but failed on the 03/02 (the third file).  Thus I think 03/03 model output grib2 file are okay.  But why the error message always appeared in the third day processing?

Also, interested things is that PMMAX and PMAVE did not generated error message when processing third file. Despite the error, the statistic seems to be the same as those created by the Perry but without the error message. My conf is similar to his. 
/gpfs/dell2/emc/verification/noscrub/Ho-Chun.Huang/metplus_aq/aqmmax
/gpfs/dell2/emc/verification/noscrub/Perry.Shafran/metplus_aq/aqmmax

run script on Venus is
/gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/scripts/run_metplus_aqmax.sh

run logout is
/gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/scripts/met_help_point_stat.log

intermediate log out are in
/gpfs/dell2/ptmp/Ho-Chun.Huang/metplus_aqmax

**Comment by georgemccabe** [comment]
Hi @Ho-ChunHuang-NOAA , it looks to me like the issue is with the value you have set for 'valid_time' in the field info.

From one of your log files (master_metplus.log.20220308200657), I found the following error:

> DEBUG 1: Forecast File: /gpfs/dell2/ptmp/Ho-Chun.Huang/com/aqm/prod/aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2
> DEBUG 1: Observation File: /gpfs/dell2/emc/verification/noscrub/Ho-Chun.Huang/metplus_aq/aqmmax/aqmmax8/prepbufr.aqm.2022030500
> .nc
> WARNING:
> WARNING: MetGrib2DataFile::data_plane_array() -> No matching records found for 'OZMAX8/L1'
> WARNING:
> DEBUG 2:
> DEBUG 2: --------------------------------------------------------------------------------
> DEBUG 2:
> DEBUG 2: Reading data for OZMAX8/L1.
> WARNING:
> WARNING: process_fcst_climo_files() -> no fields matching OZMAX8/L1 found in file: /gpfs/dell2/ptmp/Ho-Chun.Huang/com/aqm/prod/
> aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2
> WARNING:
> ERROR  :
> ERROR  : process_fcst_climo_files() -> no requested forecast data found!  Exiting...

@jprestop pulled the file in question from venus so I could take a closer look. I ran wgrib2 on the file:

**wgrib2 aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2**

> 1:0:d=2022030312:OZMAX8:1 sigma level:0-23 hour ave fcst:
> 2:124408:d=2022030312:OZMAX8:1 sigma level:24-47 hour ave fcst:

I do see the OZMAX8 field in this file. From the error in the log file, I scrolled up to see what was set for the METPLUS_FCST_FIELD environment variable:

> 03/08 20:08:15.230 metplus INFO: METPLUS_FCST_FIELD=field = [{ name="OZMAX8"; level="L1"; valid_time= "20220306_11"; }];

The valid time specified here does not match the init + forecast lead of either OZMAX8 field in this file. I was able to replicate the error by running plot_data_plane using the same field info:

**plot_data_plane aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2 out.nc 'name="OZMAX8"; level="L1"; valid_time= "20220306_11";'**

> DEBUG 1: Opening data file: /d1/personal/mccabe/met-help/1496/aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2
> WARNING:
> WARNING: MetGrib2DataFile::data_plane() -> No matching record found for 'OZMAX8/L1'
> WARNING:
> ERROR  :
> ERROR  : plot_data_plane -> trouble getting field "name="OZMAX8"; level="L1"; valid_time= "20220306_11";" from file "/d1/personal/mccabe/met-help/1496/aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2"

So while the field name was found in the file you are reading, the requested valid time for that field is not. The file has 0-23 and 24-47 forecast leads for OZMAX8, so I adjusted the valid time based on this and was able to successfully read both fields with plot_data_plane:

Initialization 20220303 12Z + 23 hour lead = 20220304_11:
**plot_data_plane aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2 out.nc 'name="OZMAX8"; level="L1"; valid_time= "20220304_11";'**

Initialization 20220303 12Z + 47 hour lead = 20220305_11:
**plot_data_plane aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2 out.nc 'name="OZMAX8"; level="L1"; valid_time= "20220305_11";'**

I asked Julie to also pull a metplus_final.conf file so I could see what was set for FCST_VAR1_OPTIONS. While this is conf not from the same run as above, I'm guessing the value you have set is similar:

`FCST_VAR1_OPTIONS = valid_time= "{valid?fmt=%Y%m%d?shift=1d}_04";set_attr_name = "PMAVE";`

Here I see that the valid time is set to the current valid time to process shifted forward in time by 1 day. I think you will want to use the actual valid time of your run to get the correct data:

`FCST_VAR1_OPTIONS = valid_time= "{valid?fmt=%Y%m%d_%H}";`

While the first few runs were running without error, it appears that they were likely using the wrong data for the analysis.

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

Thank you for your effort.

I need Perry to join the conversation.  The daily max value for average 8
hr ozone are all defined at 11Z of each day.  The daily is defined from 05Z
to 04Z.

     18 LEAD_SEQ = begin_end_incr(0,48,24)

     48 FCST_VAR1_NAME = OZMAX8
     49 FCST_VAR1_LEVELS = L1
     50 FCST_VAR1_OPTIONS = valid_time= "{valid?fmt=%Y%m%d?shift=1d}_11";

in /gpfs/dell2/emc/verification/save/Ho-Chun.Huang/METplus-4.0.0/parm/use_cases/perry/point_stat_aqmax8.conf


From your analysis, additional question is that it is a two day
verification, and why it wants to attempt lead 3 day

*#### aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2 out.nc
<http://out.nc/> 'name="OZMAX8"; level="L1"; valid_time= "20220306_11";   ?*

after
lead 1 day
*#### aqm.20220305/aqm.t12z.max_8hr_o3.148.grib2 out.nc
<http://out.nc/> 'name="OZMAX8"; level="L1"; valid_time= "20220306_11";*
lead 2 day
*#### aqm.20220304/aqm.t12z.max_8hr_o3.148.grib2 out.nc
<http://out.nc/> 'name="OZMAX8"; level="L1"; valid_time= "20220306_11";*

Did I have an error in my script/conf that forced the process to find the
lead 3 day from  aqm.20220303/aqm.t12z.max_8hr_o3.148.grib2 ?


Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Reply by georgemccabe** [reply]
For valid time 20220305 setting:

`FCST_VAR1_OPTIONS = valid_time= "{valid?fmt=%Y%m%d?shift=1d}_11";`

Will take the YYYYMMDD of the valid time (20220305) and shift it forward by 1 day to get 20220306_11. Removing `?shift=1d` will give you the current valid time you are processing.

`FCST_VAR1_OPTIONS = valid_time= "{valid?fmt=%Y%m%d}_11";`

**Reply by PerryShafran-NOAA** [reply]
Hi, @georgemccabe ,

I think that the concern for us is that if we make this change by removing the shift=1d, that will cause the other times to fail.  I think that you put that in for me in the first place when we were developing this.  Is it a mismatch in bringing in the correct grib file perhaps?

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

Also, for PMMAX and PMAVE  the processes also tried to find the lead for 3
days but did not generate the error message.  Why is the behavior different
from the OZMAX1 and OZMAX8?

Both  PMMAX and PMAVE  are defined at the _04Z of next days that is similar
to OZMAX1.


Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Reply by georgemccabe** [reply]
I would have to look at a file that contains PMMAX and PMAVE to see what records are available. I would guess that there is also a record for the 3 day lead for those fields so it is using that. It looks like the files are relatively small, so you could attach one of them to this discussion if you'd like me to take a closer look. You may have to add ".txt" to the filename to be able to upload it. Alternatively, you could upload it to our FTP.

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

I figured out that the OZMAX1 and OZMAX8 model output grib2 files only have
a two-days forecast, but PMMAX and PMAVE model output grib2 files have a
3-days forecast.  So point_stat search for the valid date for PM* will not
have an error.  For finding the valid date of "20220308_040000".

$WGRIB2 -vt aqm.20220305/aqm.t12z.max_1hr_o3.148.grib2
1:0:vt=2022030604
2:126251:vt=2022030704

$WGRIB2 -vt aqm.20220305/aqm.t12z.max_1hr_pm25.148.grib2
1:0:vt=2022030604
2:127458:vt=2022030704
3:250954:vt=2022030804

Perry and I still think the shift_1d is needed because of the way PB2NC
bufr_file created for MAX and AVE
Valid time for prepbufr file prepbufr.aqm.2022030700.nc
ozmax1   "20220308_040000" ;
ozmax8   "20220308_110000" ;
pmmax    "20220308_040000" ;
pmave    "20220308_040000" ;

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Reply by georgemccabe** [reply]
You have the shift of valid time in the FCST field info, so it will pull data from the next day for that data source. I think you want to pull the actual valid time from the FCST data since that is the valid time you want to evaluate. If the timestamp of the PB2NC output file is off, I think you would want to shift the PB2NC_OUTPUT_TEMPLATE so that it matches the valid time inside the files. That would allow you to substitute the valid time you are evaluating into the filename template for the OBS input to PointStat and get the correct data. The timing of these PrepBUFR data are often confusing to me, so I apologize if I am misunderstanding the issue.

**Reply by PerryShafran-NOAA** [reply]
@georgemccabe @Ho-ChunHuang-NOAA The prepbufr files for AQ is weird, but for the daily max/avg verification, this is what we usually have, as I put in an email to Ho-Chun yesterday:

So, looking at the verification that was valid in ops, valid day 20220308:

For ozone, the prepbufr files used were from 20220308 and 20220309
For PM, the prepbufr files used were from 2022309 and 202203010

For PM, though, the time stamp INSIDE the files I believe are 2022308 and 2022309.  That +1 is basically for the directory name that we search from, for example:

For ozone, we bring in the aqm.t12z.prepbufr.tm00 file from hourly.20220308 and hourly.20220309
For PM, we bring in the aqm.t12z.anowpm.pb.tm024 file from hourly.20220309 and hourly.20220310

But for both have the internal center date of 2022030812  and 2022030912, identical dates.

So we need to create both PB2NC and PointStat configurations to accommodate for all of this.  For the example above, since the summation is over files from 20220308 and 20220309 that the valid time would actually be 04Z on 20220309 (11Z for 8-hr ozone).  

Ho-Chun - does that look right?
George - can you offer us configurations that achieve this for us?

Perry

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

Thanks for your suggestion.  I will discuss with Perry whether we will make
an adjustment.

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, Perry:

Validate time for prepbufr.aqm.2022030700.nc (ndcump -v hdr_vld_table
prepbufr.aqm.2022030700.nc) are

ozmax1   "20220308_040000" ;
ozmax8   "20220308_110000" ;
pmmax    "20220308_040000" ;
pmave    "20220308_040000" ;

The validate time is consistent between ozone and PM despite the prepbufr
file used from different input directory.

In pb2nc_aqmax1.conf;
PB2NC_OUTPUT_TEMPLATE  = prepbufr.aqm.{valid?fmt=%Y%m%d%H}.nc

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

Please help me better understand the point-stat process.

The pb2nc file being used for the point_stat, i.e.,
prepbufr.aqm.2022030700.nc, is determined by
VALID_BEG = ${DATE}
VALID_END = ${DATE}
? From the file it finds the validate time is 20220308_04 (as for amqmax1),
correct?

Thank you for your help.

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Reply by georgemccabe** [reply]
Yes, the VALID_BEG and VALID_END values determine while valid time to substitute into the filename template for the OBS input to PointStat. See my comment below regarding reconciling the file naming so that the valid timestamp inside the PB2NC output files matches the filename.

**Comment by georgemccabe** [comment]
I'm sorry, but I'm having a hard time wrapping my head around all of this. Below is how I would approach setting this up. Hopefully that helps. If you'd like, you could send us a full set of input files (prepBUFR and forecast) files for a given valid time that you want to process along with any MET and METplus config files you use and I can use that to take a deeper dive.

If running for valid time 20220308, I would set the file window for PB2NC (OBS_PB2NC_FILE_WINDOW_BEGIN/END) to look backwards by 1 day instead of forward. This would produce an output file from PB2NC that is named with the valid time 20220308 and contain data that has a timestamp of the same day. If you must look forward with the file windows, I would shift the time in the PB2NC output template so that the time in the file matches filename, however to me that seems like it would be relevant for the next day's valid time run.

For PointStat, setting the valid_time value for the FCST data to the current valid time should read the data valid at the time you are trying to evaluate. If the OBS file created by PB2NC is named with the correct valid time in the file name, then that file could easily be read in by referring to the valid time in the OBS input template.

**Reply by georgemccabe** [reply]
Hi @Ho-ChunHuang-NOAA and @PerryShafran-NOAA , I am just following up on this discussion to see if you need additional help. Were you able to get your configuration file set up correctly?

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

I am unable to test your backward window to see if the valid time is the
same day as the filename date for PB2NC output file.

Also, for point_stat, if the fcst_input_files can not be found it prints
out an error message and lets the code continue to the next step.  If the
fcst_input_files existed and can not find the matched
variable+valid_time it is a critical error and crashes the entire batch
jobs.

Because I can not resolve the inconsistent valid time as described above, I
have to temporarily let the code not find the fcst_input_file so I can
finish the batch jobs successfully.

Any other way I can avoid the "critical error"?

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Reply by georgemccabe** [reply]
Hi @Ho-ChunHuang-NOAA, could you upload the following to the FTP:

* Log file that contains the critical error
* A full set of input files (prepBUFR and forecast) files for a given valid time that you want to process
* Any METplus and MET config files used to process these data

See [How to Send us Data](https://github.com/dtcenter/METplus/discussions/954)

**Comment by Ho-ChunHuang-NOAA** [comment]
Hi, George:

If I recall correctly, you do not have access to the WCOSS?  Would John be
able to help transfering the input data to NCAR?  Please advise.

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958

**Reply by georgemccabe** [reply]
Correct, I do not have access to WCOSS. Unfortunately the two people I typically ask to pull data from WCOSS, John and Julie, are both out on vacation this week.

**Comment by Ho-ChunHuang-NOAA** [comment, marked as answer]
Hi, George:

I can not reproduce the critical error case.  It is now running with ERROR
message without crashing my batch job.  It is possible that it is for a
different reason.  Please close this ticket.

Thanks for your attention.

Ho-Chun Huang, Ph.D.

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2792

College Park, MD 20740

***@***.*** ***@***.***>

301-683-3958
